File: host/src/sysex7.js
// 7-in-8 MSB packing — PROTOCOL.md §2.
//
// Byte-for-byte equivalent to lib/bridge_proto/sysex7.cpp. The two are
// independent implementations on purpose: host/test/roundtrip.test.js checks
// them against the same spec vector, so a change to one that the other does
// not follow shows up as a test failure rather than as corrupt serial data.

/** Bytes on the wire needed to carry `n` raw bytes. */
export function packedLen(n) {
  return n + Math.ceil(n / 7);
}

/** Largest raw length whose packed form fits in `capacity` bytes. */
export function maxRawFor(capacity) {
  const rem = capacity % 8;
  return Math.floor(capacity / 8) * 7 + (rem === 0 ? 0 : rem - 1);
}

/**
 * @param {Uint8Array} raw
 * @returns {Uint8Array} 7-bit-safe bytes
 */
export function pack(raw) {
  const out = new Uint8Array(packedLen(raw.length));
  let w = 0;
  for (let i = 0; i < raw.length; i += 7) {
    const k = Math.min(7, raw.length - i);
    const msbAt = w++;
    let msb = 0;
    for (let j = 0; j < k; j++) {
      const b = raw[i + j];
      msb |= ((b >> 7) & 1) << j;
      out[w++] = b & 0x7f;
    }
    out[msbAt] = msb;
  }
  return out;
}

export class UnpackError extends Error {
  constructor(message) {
    super(message);
    this.name = 'UnpackError';
  }
}

/**
 * Decodes 7-in-8 packed bytes. Throws UnpackError on malformed input rather
 * than guessing — a corrupt stream must be visible, not reinterpreted.
 *
 * @param {Uint8Array} packedBytes
 * @returns {Uint8Array}
 */
export function unpack(packedBytes) {
  const n = packedBytes.length;
  const out = new Uint8Array(n); // an upper bound; sliced at the end
  let w = 0;
  let i = 0;

  while (i < n) {
    const msb = packedBytes[i++];
    if (msb & 0x80) throw new UnpackError(`MSB byte 0x${msb.toString(16)} has bit 7 set`);

    const remaining = n - i;
    if (remaining === 0) throw new UnpackError('trailing MSB byte with no data');
    const k = Math.min(7, remaining);

    // The MSB byte may only claim high bits for bytes that actually follow.
    if (k < 7 && msb >> k !== 0) {
      throw new UnpackError(`MSB byte 0x${msb.toString(16)} claims more than ${k} bytes`);
    }

    for (let j = 0; j < k; j++) {
      const b = packedBytes[i + j];
      if (b & 0x80) throw new UnpackError(`data byte 0x${b.toString(16)} has bit 7 set`);
      out[w++] = b | (((msb >> j) & 1) << 7);
    }
    i += k;
  }

  return out.subarray(0, w);
}

File: lib/bridge_proto/sysex7.cpp
#include "sysex7.h"

namespace bridge {
namespace sysex7 {

size_t pack(const uint8_t* raw, size_t n, uint8_t* out) {
  size_t written = 0;
  for (size_t i = 0; i < n; i += 7) {
    const size_t k = (n - i < 7) ? (n - i) : 7;
    uint8_t* msb = &out[written++];
    *msb = 0;
    for (size_t j = 0; j < k; ++j) {
      const uint8_t b = raw[i + j];
      *msb |= static_cast<uint8_t>((b >> 7) & 1u) << j;
      out[written++] = b & 0x7F;
    }
  }
  return written;
}

UnpackResult unpack(const uint8_t* packed, size_t n, uint8_t* out,
                    size_t outCapacity) {
  size_t written = 0;
  size_t i = 0;
  while (i < n) {
    const uint8_t msb = packed[i++];
    if (msb & 0x80) return {UnpackStatus::HighBitSet, 0};

    const size_t remaining = n - i;
    if (remaining == 0) return {UnpackStatus::Truncated, 0};
    const size_t k = remaining < 7 ? remaining : 7;

    // The MSB byte may only claim high bits for bytes that actually follow.
    // A group of k bytes leaves bits k..6 of the MSB byte unused.
    if (k < 7 && (msb >> k) != 0) return {UnpackStatus::BadMsbByte, 0};

    if (written + k > outCapacity) return {UnpackStatus::OutputTooSmall, 0};
    for (size_t j = 0; j < k; ++j) {
      const uint8_t b = packed[i + j];
      if (b & 0x80) return {UnpackStatus::HighBitSet, 0};
      out[written++] = static_cast<uint8_t>(b | (((msb >> j) & 1u) << 7));
    }
    i += k;
  }
  return {UnpackStatus::Ok, written};
}

}  // namespace sysex7
}  // namespace bridge
